server/src/main/java/com/linecorp/centraldogma/server/internal/mirror/MirroringMigrationService.java

```diff
@@ -0,0 +1,428 @@
+        this.projectManager = projectManager;
+        this.commandExecutor = commandExecutor;
+    }
+
+    void migrate() throws Exception {
+
+        if (wasMigrated()) {
+            logger.debug("Mirrors and credentials have already been migrated. Skipping auto migration...");
+            return;
+        }
+
+        // Enter read-only mode.
+        commandExecutor.execute(Command.updateServerStatus(ServerStatus.REPLICATION_ONLY))
+                       .get(1, TimeUnit.MINUTES);
+        logger.info("Starting Mirrors and credentials migration ...");
+        if (commandExecutor instanceof ZooKeeperCommandExecutor) {
+            logger.debug("Waiting for 30 seconds to make sure that all cluster have been notified of the " +
+                         "read-only mode ...");
+            Thread.sleep(30000);
+        }
+
+        final Stopwatch stopwatch = Stopwatch.createStarted();
+        try {
+            for (Project project : projectManager.list().values()) {
+                logger.info("Migrating mirrors and credentials in the project: {} ...", project.name());
+                boolean processed = false;
+                final MetaRepository repository = project.metaRepo();
+                processed |= migrateCredentials(repository);
+                // Update the credential IDs in the mirrors.json file.
+                processed |= migrateMirrors(repository);
+                if (processed) {
+                    logger.info("Mirrors and credentials in the project: {} have been migrated.",
+                                project.name());
+                } else {
+                    logger.info("No legacy configurations of mirrors and credentials found in the project: {}.",
+                                project.name());
+                }
+            }
+        } catch (Exception ex) {
+            final MirrorMigrationException mirrorException = new MirrorMigrationException(
+                    "Failed to migrate mirrors and credentials. Rollback to the legacy configurations", ex);
+            try {
+                rollbackMigration();
+            } catch (Exception ex0) {
+                ex0.addSuppressed(mirrorException);
+                throw new MirrorMigrationException("Failed to rollback the mirror migration:", ex0);
+            }
+            throw mirrorException;
+        }
+
+        // Exit read-only mode.
+        commandExecutor.execute(Command.updateServerStatus(ServerStatus.WRITABLE))
+                       .get(1, TimeUnit.MINUTES);
+        logger.info("Mirrors and credentials migration has been completed. (took: {} ms.)",
+                    stopwatch.elapsed().toMillis());
+
+        shortWords = null;
+    }
+
+    private boolean wasMigrated() throws Exception {
+        for (Project project : projectManager.list().values()) {
+            final MetaRepository repository = project.metaRepo();
+            if (getMetaData(repository, PATH_LEGACY_MIRRORS_BACKUP) != null) {
```

Comment:
Question on the null check here - am I understanding correctly that if a legacy `mirrors.json`/`credentials.json` doesn't exist for a project then no backup file will exist?

i.e. If there is a repository without mirroring configured, will this method always return false?

Reply:
> if a legacy mirrors.json/credentials.json doesn't exist for a project then no backup file will exist?

Correct. no backup file won't be created if no mirror is configured.
https://github.com/line/centraldogma/blob/a80425c2780047e86ce05b04e226af08801192fd/server/src/main/java/com/linecorp/centraldogma/server/internal/mirror/MirroringMigrationService.java#L164-L167

> i.e. If there is a repository without mirroring configured, will this method always return false?

This method returns `false` but nothing will be performed.

Reply:
> This method returns false but nothing will be performed.

I see. I just wanted to make sure the following lines won't be executed after a migration is successfully completed.

https://github.com/line/centraldogma/blob/f0184f023f35208aa91949502439e7e3824a9a40/server/src/main/java/com/linecorp/centraldogma/server/internal/mirror/MirroringMigrationService.java#L100-L108

Reply:
`/mirror-migration-job.json` is added to check if a migration has been performed before.
`/mirror-migration-job.json` will be written even if there are no mirror and credential configurations.